File: lib-segment/src/main/java/websiteschema/mpsegment/filter/UnknownPlaceFilter.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package websiteschema.mpsegment.filter;

import websiteschema.mpsegment.dict.POSUtil;

/**
 *
 * @author ray
 */
public class UnknownPlaceFilter extends AbstractSegmentFilter {

    private static String adminLevels = "省市县区乡镇村旗州";
    private boolean recognizeDiMing = true;

    @Override
    public void doFilter() {
        if (recognizeDiMing) {
            int length = segmentResult.length();
            for (int wordI = 0; wordI < length; wordI++) {
                if (wordPosIndexes[wordI] <= 0) {
                    int pos = segmentResult.getPOS(wordI);
                    if (pos == POSUtil.POS_NS && wordI + 1 < length) {
                        String word = segmentResult.getWord(wordI);
                        if (segmentResult.getWord(wordI + 1).length() == 1 && adminLevels.indexOf(segmentResult.getWord(wordI + 1)) > 0 && word.lastIndexOf(segmentResult.getWord(wordI + 1)) != length - 1) {
                            mergeWordsWithPOS(wordI, wordI + 1, POSUtil.POS_NS);
                        }
                    }
                }
            }
        }
    }
}

File: lib-segment/src/main/scala/me/yingrui/segment/filter/UnknownPlaceFilter.scala
/*
* To change this template, choose Tools | Templates
* and open the template in the editor.
*/
package me.yingrui.segment.filter

import me.yingrui.segment.dict.POSUtil

/**
*
* @author ray
*/
class UnknownPlaceFilter extends AbstractSegmentFilter {

    private val adminLevels = "省市县区乡镇村旗州"
    private val recognizeDiMing : Boolean = true

    override def doFilter() {
        if (recognizeDiMing) {
            val length = segmentResult.length()
            for (wordI <- 0 until length) {
                if (isNotMarked(wordI)) {
                    val pos = segmentResult.getPOS(wordI)
                    if (pos == POSUtil.POS_NS && wordI + 1 < length) {
                        val word = segmentResult.getWord(wordI)
                        if (segmentResult.getWord(wordI + 1).length() == 1 && adminLevels.indexOf(segmentResult.getWord(wordI + 1)) > 0 && word.lastIndexOf(segmentResult.getWord(wordI + 1)) != length - 1) {
                            setWordIndexesAndPOSForMerge(wordI, wordI + 1, POSUtil.POS_NS)
                        }
                    }
                }
            }
        }
    }
}
